designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala

```diff
@@ -186,72 +97,65 @@ class DefaultComponentService private(config: Config,
+    processService
+      .getProcesses[DisplayableProcess](loggedUser)
+      .map(_.filter(p => categories.contains(p.processCategory))) //TODO: move it to service?
+      .map(processes => ComponentsUsageHelper.computeComponentsUsageCount(componentIdProvider, processes))
+  }
+
   private def extractUserComponentsFromProcessingType(processingTypeData: ProcessingTypeData,
-                                                      processingType: String,
+                                                      processingType: ProcessingType,
                                                       user: LoggedUser,
                                                       componentUsages: Map[ComponentId, Long]): Future[List[ComponentListElement]] = {
     processService
       .getSubProcesses(processingTypes = Some(List(processingType)))(user)
-      .map(subprocesses => {
-        /**
-          * TODO: Right now we use UIProcessObjectsFactory for extract components data, because there is assigned logic
-          * responsible for: hiding, mapping group name, etc.. We should move this logic to another place, because
-          * UIProcessObjectsFactory does many other things, things that we don't need here..
-          */
-        val uiProcessObjects = UIProcessObjectsFactory.prepareUIProcessObjects(
-          processingTypeData.modelData,
-          processingTypeData.deploymentManager,
-          processingTypeData.typeSpecificInitialData,
-          user,
-          subprocesses,
-          isSubprocess = false, //It excludes fragment's components: input / output
-          categoryService,
-          processingTypeData.additionalPropertiesConfig,
-          processingType
-        )
-
-        def createLinks(componentId: ComponentId, component: ComponentTemplate): List[ComponentLink] = {
-          val componentLinks = componentLinksConfig
-            .filter(_.isAvailable(component.`type`))
-            .map(_.toComponentLink(componentId, component.label))
-
-          //If component configuration contains documentation link then we add base link
-          getComponentDoc(uiProcessObjects.componentsConfig, component)
-            .map(ComponentLink.createDocumentationLink)
-            .map(doc => List(doc) ++ componentLinks)
-            .getOrElse(componentLinks)
-        }
-
-        uiProcessObjects
-          .componentGroups
-          .flatMap(group => group.components.map(com => {
-            val componentId = componentIdProvider.createComponentId(processingType, com.label, com.`type`)
-            val icon = getComponentIcon(uiProcessObjects.componentsConfig, com)
-            val links = createLinks(componentId, com)
-            val usageCount = componentUsages.getOrElse(componentId, 0L)
+      .map { subprocesses =>
+        val componentObjects = componentObjectsService.prepare(processingType, processingTypeData, user, subprocesses)
```

Comment:
I think the assumption is that a fragment never gets a component id set via configuration. Am I right @lciolecki ?

Reply:
This assumption is fine for me, I don't see problem with `ComponentObjectsService.prepareWithoutFragments`, I see a problem with `ComponentObjectsService.prepare`. IMO we need to write comment why we don't need to validate this set of data. It is not clear because:
- in `CombinedProcessingTypeData` there is only lazy concept of `ComponentIdProvider` available. Why to assume that this provider return not colliding component ids for any set of components?
- During preparation of  `ComponentIdProvider` we do validations only on set with fragment excludes. Why fragments won't collide with this set?

I think that we should add two comments:
1. In CombinedProcessingTypeData, next to ComponentIdProvider
2. Next to componentObjectsService.prepare invocation  - about assumption that fragments has ids in separate namespaces: $processingType-fragments- and because of that we don't need to validate against collisions